mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
+                return 1
+            fi
+
+            # start cherry-picking with latest merge, squash everything before
+            local f_first_pick_base=${f_latest_branch_point_commit}
+            local f_latest_merge_commit=$(git log --merges --format='%H' --ancestry-path -1 ${f_latest_branch_point_commit}..${f_mainline_merge}^2)
+            if [ -n "${f_latest_merge_commit}" ]; then
+                echo "Cherry-picking squashed k8s.io/kubernetes commits $(kube-commit ${f_latest_branch_point_commit})..$(kube-commit ${f_latest_merge_commit}) because the last one is a merge: $(git show --format="%s" -q ${f_latest_merge_commit})"
+
+                # reset Godeps.json?
+                if [ -n "$(godep-changes ${f_latest_branch_point_commit} ${f_latest_merge_commit})" ]; then
+                    reset-godeps ${f_latest_branch_point_commit}
+                    dst_needs_godeps_update=true
+                fi
+
+                if ! git diff ${f_latest_branch_point_commit} ${f_latest_merge_commit} | git apply --index; then
+                    echo
+                    show-working-dir-status
+                    return 1
+                fi
+                git-with-user commit -q -m "sync: squashed up to merge $(kube-commit ${f_latest_merge_commit}) in $(kube-commit ${f_mainline_merge})"
+                ensure-clean-working-dir
+
+                # we start cherry-picking now from f_latest_merge_commit up to the actual Github merge into the mainline
+                f_first_pick_base=${f_latest_merge_commit}
+            fi
+            for f_commit in $(git log --format='%H' --reverse ${f_first_pick_base}..${f_mainline_merge}^2); do
+                # reset Godeps.json?
+                if [ -n "$(godep-changes ${f_commit})" ]; then
+                    reset-godeps ${f_commit}^
+                    dst_needs_godeps_update=true
+                fi
+
+                echo "Cherry-picking k8s.io/kubernetes commit ${f_commit}: $(git show --format="%s" -q ${f_commit})."
+                if ! git-with-user cherry-pick --keep-redundant-commits ${f_commit} >/dev/null; then
+                    echo
+                    show-working-dir-status
+                    return 1
+                fi
+                ensure-clean-working-dir
+            done
+
+            # commit empty PR merge. This will carry the actual SHA1 from the upstream commit. It will matchs tags as well.
+            echo "Cherry-picking k8s.io/kubernetes merge  $(kube-commit ${f_mainline_merge}): $(git show --format="%s" -q ${f_mainline_merge})."
+            local date=$(git show --format="%aD" -q ${f_mainline_merge}) # author and committer date is equal for PR merges
+            git reset -q $(GIT_COMMITTER_DATE="${date}" GIT_AUTHOR_DATE="${date}" git-with-user commit-tree -p ${dst_merge_point_commit} -p HEAD -m "$(git show --format="%B" -q ${f_mainline_merge})" HEAD^{tree})
```

Comment:
How do you guarantee this comit "carry actual SHA1 from the upstream commit"?

Reply:
The commit msg comes from `f_mainline_merge` which has the `Kubernetes-commit: ...` annotation.